# Project 4
# Francesca Lama

In [506]:
%matplotlib notebook

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
pd.set_option('display.max_columns', None)

In [507]:
import warnings
warnings.filterwarnings('ignore')

## Question 1
### 1 a
Here I imported the data into Jupyter notebook, subsetted it and then concattonated it into one wide sample. Then I cleaned the data and reshaped it into long data. I am still focusing on the two business groups as I was in project 3 which are Grocery/Convenience stores and Restaurants/Bars. 

In [508]:
## importing 4 months of data
# data1 = pd.read_csv('Project_4_data/data01.csv', 
#                    dtype = {'poi_cbg':'object','naics_code': 'object'})

In [509]:
# data2 = pd.read_csv('Project_4_data/data04.csv', 
#                    dtype = {'poi_cbg':'object','naics_code': 'object'})

In [510]:
# data3 = pd.read_csv('Project_4_data/data07.csv', 
#                    dtype = {'poi_cbg':'object','naics_code': 'object'})

In [511]:
# data4 = pd.read_csv('Project_4_data/data10.csv', 
#                    dtype = {'poi_cbg':'object','naics_code': 'object'})

In [512]:
## combining the data 
# data = pd.concat([data1, data2, data3, data4], ignore_index = True)

In [513]:
# data.head()

In [514]:
## subsetting the data into a smaller sample (Phoenix, AZ)
# az = data[data['region'] == 'AZ']

In [515]:
# az.head()

In [516]:
# phoenix = az[az['city'] == 'Phoenix']

In [517]:
# phoenix.head()

In [518]:
# len(phoenix)

In [519]:
## deleting missing data
# phoenix = phoenix[~phoenix.isna().any(axis = 1)].copy()

In [520]:
## define bizgroups
## grocery store/convenience stores
# phoenix.loc[phoenix['naics_code'].str.startswith('4451'), 'bizgroup'] = '1'
## food service (restaurants and bars)
# phoenix.loc[phoenix['naics_code'].str.startswith('722'), 'bizgroup'] = '2'

In [521]:
## 446110, 721211, and 721120 should not be there
# phoenix.naics_code.unique()

In [522]:
## data cleaning
# phoenix['visits_by_day'] = phoenix['visits_by_day'].str.replace('\[|\]', '')

# cols = []
# for i in range (31):
#    v = 'dailyvisits{}'.format(i)
#    cols.append(v)
# phoenix[cols] = phoenix['visits_by_day'].str.split(',', expand = True)

# phoenix['id'] = phoenix.index
# long = pd.wide_to_long(phoenix, ['dailyvisits'], i = 'id', j = 'dayofmonth')

# long = long[~long['dailyvisits'].isnull()].copy()
# long.reset_index(inplace = True)

# long['date'] = pd.to_datetime(long['date_range_start']) \
#                + pd.to_timedelta(long['dayofmonth'],
#                      unit = 'd')

# long['dayofweek'] = long['date'].dt.dayofweek
# daynames = ['Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat', 'Sun']
# long['dayname'] = long['dayofweek'].replace([0,1,2,3,4,5,6], daynames)
# long['dayname'] = pd.Categorical(long['dayname'], ordered = True,
#                                 categories = daynames)
# long['weekend'] = np.where(long['dayname'].isin(['Sat', 'Sun']), 1, 0)

# long.drop(columns = 'visits_by_day', inplace = True)

In [523]:
## saving data 
# long.to_csv('clean_data/clean.csv', index = False)

In [524]:
# view of data
phoenix = pd.read_csv('clean_data/clean.csv',
               dtype = {'poi_cbg':'object','naics_code':'object'})
phoenix.head()

id  dayofmonth region     city       poi_cbg date_range_start naics_code  \
0  149           0     AZ  Phoenix  040131124011       2020-01-01     722511   
1  169           0     AZ  Phoenix  040131059001       2020-01-01     445110   
2  288           0     AZ  Phoenix  040131168001       2020-01-01     722511   
3  469           0     AZ  Phoenix  040131063002       2020-01-01     722511   
4  641           0     AZ  Phoenix  040131114012       2020-01-01     722511   

   bizgroup                   safegraph_place_id  dailyvisits        date  \
0       2.0  sg:51b32e81b2e04a6ba7f01770ea911d70           22  2020-01-01   
1       1.0  sg:5c2b32869cbe482d881c0082a3d037bf            1  2020-01-01   
2       2.0  sg:a4a397b4ccf14f9cbb2c608565ae31c0            2  2020-01-01   
3       2.0  sg:2ba8d5d7f63a4638b3578a064515a815            1  2020-01-01   
4       2.0  sg:88bfaf9c4379486e9f610b55ffdba1d6            7  2020-01-01   

   dayofweek dayname  weekend  
0          2     Wed        0  
1          2     Wed        0  
2          2     Wed        0  
3          2     Wed        0  
4          2     Wed        0

### 1 b
I started by importing the cbg data, and I stuck with the variables that were already in the data set. I dropped the unmatched observations. I merged the data and named it 'LocalSafegraph'.
* There are 252,856 observations
* There are 2,362 unique places 
* There are 3 unique NAICS codes
* There are 196,979 unique places for restaurants, 29,181 unique places for Grocery stores, and 26,696 unique places for Convenience stores
* The average population for each observation is 1632.27
There is also only one value for the variable FIPS_5 variable I created and that value is 04013.

In [525]:
# Census data we use, including the match id 'census_block_group' and 16 variables
cbg = pd.read_csv('clean_data/cbg.csv',
    dtype = {'census_block_group':'object'})
cbg.head()

census_block_group  B01001e1  B01001e2  B01002e1  B01002e2  B01002e3  \
0       010010201001       745       356      34.1      37.1      33.7   
1       010010201002      1265       639      41.8      39.9      44.8   
2       010010202001       960       534      38.2      23.6      45.0   
3       010010202002      1236       634      39.7      33.5      43.3   
4       010010203001      2364      1125      34.9      34.1      39.3   

   B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  B11001e3  B11001e4  \
0       585       160         0       284       193       112        81   
1      1083       104         9       456       390       310        80   
2       361       568         0       386       244       145        99   
3       615       571        24       452       286       184       102   
4      1481       515        27       824       560       435       125   

   B11001e5  B11001e6  B11001e7  B11001e8  B11001e9  B15003e1  B15003e2  \
0        11        70        91        63        28       474         0   
1         4        76        66        58         8       824         7   
2         6        93       142       133         9       602         9   
3         9        93       166       153        13       861         7   
4         0       125       264       228        36      1648        46   

   B15003e3  B15003e4  B15003e5  B15003e6  B15003e7  B15003e8  B15003e9  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         4        15   
3         0         0         0         0         0         0         0   
4         0         0        18         0         0         0         0   

   B15003e10  B15003e11  B15003e12  B15003e13  B15003e14  B15003e15  \
0         23          0          0         53          6         10   
1          0          0          0         27         10         11   
2          0          0          6         17          4         63   
3          6          6         16         18         48         28   
4          0          5          5         39         42         33   

   B15003e16  B15003e17  B15003e18  B15003e19  B15003e20  B15003e21  \
0         11        128         21         13         88         15   
1          9        216         51         21        137         34   
2          3        211         51         70         64         15   
3         24        255         59         63         90         74   
4         20        536         40        122        264        112   

   B15003e22  B15003e23  B15003e24  B15003e25  B19013e1    B19025e1  B19301e1  
0         60         23          7         16       NaN  14126600.0   20365.0  
1        156        118         18          9   77813.0  41514300.0   33336.0  
2         54          7          9          0   25179.0  15608200.0   17047.0  
3        113         39          9          6   45104.0  25847500.0   21400.0  
4        205        133         10         18   55222.0  50827900.0   23106.0

In [526]:
# I include two variables that I create
# B15003e2-18: high school and below
# B15003e19-25: college and higher
cbg['EDU_under'] = cbg.loc[:,'B15003e2':'B15003e16'].sum(axis=1)/cbg['B15003e1']
cbg['EDU_higher'] = cbg.loc[:,'B15003e21':'B15003e25'].sum(axis=1)/cbg['B15003e1']

cbg1 = cbg.loc[:,['census_block_group','B01001e1','B01001e2','B01002e1',
                  'B01002e2','B01002e3','B02001e2','B02001e3','B02001e5',
                  'B11001e1','B11001e2','B11001e3','EDU_under','EDU_higher',
                  'B19013e1','B19025e1','B19301e1']]
cbg1.head()

census_block_group  B01001e1  B01001e2  B01002e1  B01002e2  B01002e3  \
0       010010201001       745       356      34.1      37.1      33.7   
1       010010201002      1265       639      41.8      39.9      44.8   
2       010010202001       960       534      38.2      23.6      45.0   
3       010010202002      1236       634      39.7      33.5      43.3   
4       010010203001      2364      1125      34.9      34.1      39.3   

   B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  B11001e3  EDU_under  \
0       585       160         0       284       193       112   0.217300   
1      1083       104         9       456       390       310   0.077670   
2       361       568         0       386       244       145   0.200997   
3       615       571        24       452       286       184   0.177700   
4      1481       515        27       824       560       435   0.126214   

   EDU_higher  B19013e1    B19025e1  B19301e1  
0    0.255274       NaN  14126600.0   20365.0  
1    0.406553   77813.0  41514300.0   33336.0  
2    0.141196   25179.0  15608200.0   17047.0  
3    0.279907   45104.0  25847500.0   21400.0  
4    0.290049   55222.0  50827900.0   23106.0

In [527]:
# merge data
LocalSafegraph = pd.merge(phoenix, cbg1, left_on = 'poi_cbg', right_on = 'census_block_group', how = 'left')
LocalSafegraph.drop(columns = 'census_block_group', inplace = True)

In [528]:
# double-check: all rows in df should be merged into df2
print(len(phoenix) - len(LocalSafegraph))

0


In [529]:
# drop all rows with missing values
LocalSafegraph = LocalSafegraph[~LocalSafegraph.isna().any(axis=1)].copy()

In [530]:
LocalSafegraph.head()

id  dayofmonth region     city       poi_cbg date_range_start naics_code  \
0  149           0     AZ  Phoenix  040131124011       2020-01-01     722511   
1  169           0     AZ  Phoenix  040131059001       2020-01-01     445110   
2  288           0     AZ  Phoenix  040131168001       2020-01-01     722511   
3  469           0     AZ  Phoenix  040131063002       2020-01-01     722511   
4  641           0     AZ  Phoenix  040131114012       2020-01-01     722511   

   bizgroup                   safegraph_place_id  dailyvisits        date  \
0       2.0  sg:51b32e81b2e04a6ba7f01770ea911d70           22  2020-01-01   
1       1.0  sg:5c2b32869cbe482d881c0082a3d037bf            1  2020-01-01   
2       2.0  sg:a4a397b4ccf14f9cbb2c608565ae31c0            2  2020-01-01   
3       2.0  sg:2ba8d5d7f63a4638b3578a064515a815            1  2020-01-01   
4       2.0  sg:88bfaf9c4379486e9f610b55ffdba1d6            7  2020-01-01   

   dayofweek dayname  weekend  B01001e1  B01001e2  B01002e1  B01002e2  \
0          2     Wed        0      1605       713      22.0      20.3   
1          2     Wed        0      1276       707      46.4      43.7   
2          2     Wed        0      1066       543      25.5      36.9   
3          2     Wed        0      1551       730      48.4      48.3   
4          2     Wed        0      2065       979      31.8      31.1   

   B01002e3  B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  B11001e3  \
0      30.1       806       159        22       475       289        59   
1      49.6      1069        63        36       525       234       153   
2      20.0       975        29         0       315       217        79   
3      54.1      1504        24         0       627       273       273   
4      32.7      1143       120         7       610       456       190   

   EDU_under  EDU_higher  B19013e1    B19025e1  B19301e1  
0   0.561995    0.048518   18972.0  10545000.0    7621.0  
1   0.057432    0.247748   27298.0  22996600.0   19414.0  
2   0.545966    0.105066   13250.0   8022300.0    8608.0  
3   0.029752    0.582645   57098.0  54859600.0   38017.0  
4   0.403846    0.173077   38839.0  27257000.0   15326.0

In [531]:
# checking observations 
len(LocalSafegraph)

252856

In [532]:
# unique places
unique_place = LocalSafegraph.safegraph_place_id.unique()
len(unique_place)

2362

In [533]:
# unique NAICS codes
unique_codes = LocalSafegraph.naics_code.unique()
len(unique_codes)

3

In [534]:
# list of unique codes
# 445120 and 445110 represent Grocery/Convenience stores 
# 722511 represents  full service restaurants 
LocalSafegraph.naics_code.unique()

array(['722511', '445110', '445120'], dtype=object)

In [535]:
# number of unique places for each naics code
LocalSafegraph['naics_code'].value_counts()

722511    196979
445110     29181
445120     26696
Name: naics_code, dtype: int64

In [536]:
# average population for each observation
LocalSafegraph['B01001e1'].mean(axis=0)

1632.2722854114595

In [537]:
# creating FIPS_5digits 
LocalSafegraph['FIPS_5'] = LocalSafegraph['poi_cbg'].str[0:5]
print(LocalSafegraph['FIPS_5'])

0         04013
1         04013
2         04013
3         04013
4         04013
          ...  
273567    04013
273568    04013
273569    04013
273570    04013
273571    04013
Name: FIPS_5, Length: 252856, dtype: object


In [538]:
# checking FIPS_5 unique values
LocalSafegraph['FIPS_5'].nunique()

1

In [539]:
LocalSafegraph['month'] = pd.to_datetime(LocalSafegraph['date']).dt.month.astype('str')
LocalSafegraph['bizgroup'] = LocalSafegraph['bizgroup'].astype('str')

### 1c
I started by importing the data and only keeping the relevant observations with the FIPS_5 code (4013).
* There are 428 observations in this dataset
* The minimum date is 01/26/2020 and the maximum dates is 03/28/2021
* The average number of daily cases is 166545.36
* The average number of daily deaths is 3048.94

In [540]:
# importing data
COVID = pd.read_csv('clean_data/us-counties.csv',
    dtype = {'fips':'object'})

In [541]:
# showing covid cases from Maricopa county 
# also shows min and max dates
azCovid = COVID[COVID['fips'] == '04013']
azCovid

date    county    state   fips   cases  deaths
8        2020-01-26  Maricopa  Arizona  04013       1     0.0
13       2020-01-27  Maricopa  Arizona  04013       1     0.0
18       2020-01-28  Maricopa  Arizona  04013       1     0.0
23       2020-01-29  Maricopa  Arizona  04013       1     0.0
28       2020-01-30  Maricopa  Arizona  04013       1     0.0
...             ...       ...      ...    ...     ...     ...
1150992  2021-03-24  Maricopa  Arizona  04013  522323  9601.0
1154239  2021-03-25  Maricopa  Arizona  04013  522285  9619.0
1157486  2021-03-26  Maricopa  Arizona  04013  522624  9630.0
1160734  2021-03-27  Maricopa  Arizona  04013  523088  9640.0
1163982  2021-03-28  Maricopa  Arizona  04013  523424  9644.0

[428 rows x 6 columns]

In [542]:
# number of observations
len(azCovid)

428

In [543]:
# average daily cases
azCovid['cases'].mean(axis=0)

166545.3621495327

In [544]:
# average daily deaths
azCovid['deaths'].mean(axis=0)

3048.9439252336447

### 1d
* There are 366 observations in this dataset
* The average air temperature is 77 degrees
* The average precipitation is 0.013661

In [545]:
azWeather = pd.read_csv('clean_data/weather.csv')

In [546]:
# number of observations
len(azWeather)

366

In [547]:
azWeather.head()

STATION                    NAME        DATE  PRCP  SNOW  SNWD  TAVG  \
0  USW00023183  PHOENIX AIRPORT, AZ US  2020-01-01   0.0   0.0   0.0  50.0   
1  USW00023183  PHOENIX AIRPORT, AZ US  2020-01-02   0.0   0.0   0.0  51.0   
2  USW00023183  PHOENIX AIRPORT, AZ US  2020-01-03   0.0   0.0   0.0  52.0   
3  USW00023183  PHOENIX AIRPORT, AZ US  2020-01-04   0.0   0.0   0.0  53.0   
4  USW00023183  PHOENIX AIRPORT, AZ US  2020-01-05   0.0   0.0   0.0  54.0   

   TMAX  TMIN  
0    60    40  
1    64    41  
2    64    41  
3    68    41  
4    70    42

In [548]:
# average air temp
azWeather['TAVG'].mean(axis=0)

77.23691460055096

In [549]:
# average precipitation
azWeather['PRCP'].mean(axis=0)

0.01366120218579235

## Question 2
### 2a
There are 52,800 missing values in the daily count of COVID cases and COVID deaths. I proceeded to fill these missing values with 0. After this, the average daily COVID cases and daily COVID deaths in azMerged is 57,678.7 and 1,169.9 respectively. 
I then checked to see if any dates had missing values for air temperature and precipitation. I deleted the dates that did contain those missing values and now there are 247,068 observations in LocalMerged as opposed to 252,856.

In [550]:
# merging the data
Merge1 = pd.merge(LocalSafegraph, azCovid, left_on = 'date', right_on = 'date', how = 'left')
azMerged = pd.merge(Merge1, azWeather, left_on = 'date', right_on = 'DATE', how = 'left')

In [551]:
azMerged

id  dayofmonth region     city       poi_cbg date_range_start  \
0           149           0     AZ  Phoenix  040131124011       2020-01-01   
1           169           0     AZ  Phoenix  040131059001       2020-01-01   
2           288           0     AZ  Phoenix  040131168001       2020-01-01   
3           469           0     AZ  Phoenix  040131063002       2020-01-01   
4           641           0     AZ  Phoenix  040131114012       2020-01-01   
...         ...         ...    ...      ...           ...              ...   
252851  2318912          30     AZ  Phoenix  040136114002       2020-10-01   
252852  2319143          30     AZ  Phoenix  040131132021       2020-10-01   
252853  2319673          30     AZ  Phoenix  040131107021       2020-10-01   
252854  2319760          30     AZ  Phoenix  040131132022       2020-10-01   
252855  2319924          30     AZ  Phoenix  040131138012       2020-10-01   

       naics_code bizgroup                   safegraph_place_id  dailyvisits  \
0          722511      2.0  sg:51b32e81b2e04a6ba7f01770ea911d70           22   
1          445110      1.0  sg:5c2b32869cbe482d881c0082a3d037bf            1   
2          722511      2.0  sg:a4a397b4ccf14f9cbb2c608565ae31c0            2   
3          722511      2.0  sg:2ba8d5d7f63a4638b3578a064515a815            1   
4          722511      2.0  sg:88bfaf9c4379486e9f610b55ffdba1d6            7   
...           ...      ...                                  ...          ...   
252851     722511      2.0  sg:b3cc50b44f83489392ef84d04b1f8285           20   
252852     722511      2.0  sg:6f840b686e66467cad1121bda3e1904e            0   
252853     722511      2.0  sg:14a0d9c08e2f43798e8c0efd8050a2b6            5   
252854     722511      2.0  sg:568bf3f96ace415d8dae1046436b3508            5   
252855     722511      2.0  sg:fdab0927b0c54d9caefcac61d6e0c9b9            1   

              date  dayofweek dayname  weekend  B01001e1  B01001e2  B01002e1  \
0       2020-01-01          2     Wed        0      1605       713      22.0   
1       2020-01-01          2     Wed        0      1276       707      46.4   
2       2020-01-01          2     Wed        0      1066       543      25.5   
3       2020-01-01          2     Wed        0      1551       730      48.4   
4       2020-01-01          2     Wed        0      2065       979      31.8   
...            ...        ...     ...      ...       ...       ...       ...   
252851  2020-10-31          5     Sat        1      1780      1074      45.1   
252852  2020-10-31          5     Sat        1      1022       532      28.7   
252853  2020-10-31          5     Sat        1      2350      1229      35.0   
252854  2020-10-31          5     Sat        1      1107       717      29.0   
252855  2020-10-31          5     Sat        1       734       399      24.2   

        B01002e2  B01002e3  B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  \
0           20.3      30.1       806       159        22       475       289   
1           43.7      49.6      1069        63        36       525       234   
2           36.9      20.0       975        29         0       315       217   
3           48.3      54.1      1504        24         0       627       273   
4           31.1      32.7      1143       120         7       610       456   
...          ...       ...       ...       ...       ...       ...       ...   
252851      42.6      48.4      1747         0         0       616       514   
252852      28.2      28.9       559        55         0       326       177   
252853      34.6      36.1      1758        20       116       683       471   
252854      28.8      29.3       342        90         4       350       240   
252855      24.5      21.9       336        16         0       194       116   

        B11001e3  EDU_under  EDU_higher  B19013e1    B19025e1  B19301e1  \
0             59   0.561995    0.048518   18972.0  10545000.0    7621.0   
1            153   0.057432    0.247748   2

In [552]:
# number of missing values for cases
azMerged['cases'].isna().sum()

52800

In [553]:
# number of missing values for deaths
azMerged['deaths'].isna().sum()

52800

In [554]:
# replacing missing values with 0
azMerged['cases'] = azMerged['cases'].fillna(0)
azMerged['deaths'] = azMerged['deaths'].fillna(0)

In [555]:
azMerged

id  dayofmonth region     city       poi_cbg date_range_start  \
0           149           0     AZ  Phoenix  040131124011       2020-01-01   
1           169           0     AZ  Phoenix  040131059001       2020-01-01   
2           288           0     AZ  Phoenix  040131168001       2020-01-01   
3           469           0     AZ  Phoenix  040131063002       2020-01-01   
4           641           0     AZ  Phoenix  040131114012       2020-01-01   
...         ...         ...    ...      ...           ...              ...   
252851  2318912          30     AZ  Phoenix  040136114002       2020-10-01   
252852  2319143          30     AZ  Phoenix  040131132021       2020-10-01   
252853  2319673          30     AZ  Phoenix  040131107021       2020-10-01   
252854  2319760          30     AZ  Phoenix  040131132022       2020-10-01   
252855  2319924          30     AZ  Phoenix  040131138012       2020-10-01   

       naics_code bizgroup                   safegraph_place_id  dailyvisits  \
0          722511      2.0  sg:51b32e81b2e04a6ba7f01770ea911d70           22   
1          445110      1.0  sg:5c2b32869cbe482d881c0082a3d037bf            1   
2          722511      2.0  sg:a4a397b4ccf14f9cbb2c608565ae31c0            2   
3          722511      2.0  sg:2ba8d5d7f63a4638b3578a064515a815            1   
4          722511      2.0  sg:88bfaf9c4379486e9f610b55ffdba1d6            7   
...           ...      ...                                  ...          ...   
252851     722511      2.0  sg:b3cc50b44f83489392ef84d04b1f8285           20   
252852     722511      2.0  sg:6f840b686e66467cad1121bda3e1904e            0   
252853     722511      2.0  sg:14a0d9c08e2f43798e8c0efd8050a2b6            5   
252854     722511      2.0  sg:568bf3f96ace415d8dae1046436b3508            5   
252855     722511      2.0  sg:fdab0927b0c54d9caefcac61d6e0c9b9            1   

              date  dayofweek dayname  weekend  B01001e1  B01001e2  B01002e1  \
0       2020-01-01          2     Wed        0      1605       713      22.0   
1       2020-01-01          2     Wed        0      1276       707      46.4   
2       2020-01-01          2     Wed        0      1066       543      25.5   
3       2020-01-01          2     Wed        0      1551       730      48.4   
4       2020-01-01          2     Wed        0      2065       979      31.8   
...            ...        ...     ...      ...       ...       ...       ...   
252851  2020-10-31          5     Sat        1      1780      1074      45.1   
252852  2020-10-31          5     Sat        1      1022       532      28.7   
252853  2020-10-31          5     Sat        1      2350      1229      35.0   
252854  2020-10-31          5     Sat        1      1107       717      29.0   
252855  2020-10-31          5     Sat        1       734       399      24.2   

        B01002e2  B01002e3  B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  \
0           20.3      30.1       806       159        22       475       289   
1           43.7      49.6      1069        63        36       525       234   
2           36.9      20.0       975        29         0       315       217   
3           48.3      54.1      1504        24         0       627       273   
4           31.1      32.7      1143       120         7       610       456   
...          ...       ...       ...       ...       ...       ...       ...   
252851      42.6      48.4      1747         0         0       616       514   
252852      28.2      28.9       559        55         0       326       177   
252853      34.6      36.1      1758        20       116       683       471   
252854      28.8      29.3       342        90         4       350       240   
252855      24.5      21.9       336        16         0       194       116   

        B11001e3  EDU_under  EDU_higher  B19013e1    B19025e1  B19301e1  \
0             59   0.561995    0.048518   18972.0  10545000.0    7621.0   
1            153   0.057432    0.247748   2

In [556]:
# average daily cases
azMerged['cases'].mean(axis=0)

57678.70235232702

In [557]:
# average daily deaths
azMerged['deaths'].mean(axis=0)

1169.9024504065555

In [558]:
# checking for missing values in air temp
azMerged['TAVG'].isna().values.any()

True

In [559]:
# checking for missing values in precipitation
azMerged['PRCP'].isna().values.any()

False

In [560]:
#dropping dates that contain missing values in air temp and precipitation
azMerged = azMerged.dropna(subset=['TAVG', 'PRCP'])

In [561]:
# deleting unnecessary columns
azMerged.drop(['SNOW', 'SNWD'], axis = 1)

id  dayofmonth region     city       poi_cbg date_range_start  \
0           149           0     AZ  Phoenix  040131124011       2020-01-01   
1           169           0     AZ  Phoenix  040131059001       2020-01-01   
2           288           0     AZ  Phoenix  040131168001       2020-01-01   
3           469           0     AZ  Phoenix  040131063002       2020-01-01   
4           641           0     AZ  Phoenix  040131114012       2020-01-01   
...         ...         ...    ...      ...           ...              ...   
252851  2318912          30     AZ  Phoenix  040136114002       2020-10-01   
252852  2319143          30     AZ  Phoenix  040131132021       2020-10-01   
252853  2319673          30     AZ  Phoenix  040131107021       2020-10-01   
252854  2319760          30     AZ  Phoenix  040131132022       2020-10-01   
252855  2319924          30     AZ  Phoenix  040131138012       2020-10-01   

       naics_code bizgroup                   safegraph_place_id  dailyvisits  \
0          722511      2.0  sg:51b32e81b2e04a6ba7f01770ea911d70           22   
1          445110      1.0  sg:5c2b32869cbe482d881c0082a3d037bf            1   
2          722511      2.0  sg:a4a397b4ccf14f9cbb2c608565ae31c0            2   
3          722511      2.0  sg:2ba8d5d7f63a4638b3578a064515a815            1   
4          722511      2.0  sg:88bfaf9c4379486e9f610b55ffdba1d6            7   
...           ...      ...                                  ...          ...   
252851     722511      2.0  sg:b3cc50b44f83489392ef84d04b1f8285           20   
252852     722511      2.0  sg:6f840b686e66467cad1121bda3e1904e            0   
252853     722511      2.0  sg:14a0d9c08e2f43798e8c0efd8050a2b6            5   
252854     722511      2.0  sg:568bf3f96ace415d8dae1046436b3508            5   
252855     722511      2.0  sg:fdab0927b0c54d9caefcac61d6e0c9b9            1   

              date  dayofweek dayname  weekend  B01001e1  B01001e2  B01002e1  \
0       2020-01-01          2     Wed        0      1605       713      22.0   
1       2020-01-01          2     Wed        0      1276       707      46.4   
2       2020-01-01          2     Wed        0      1066       543      25.5   
3       2020-01-01          2     Wed        0      1551       730      48.4   
4       2020-01-01          2     Wed        0      2065       979      31.8   
...            ...        ...     ...      ...       ...       ...       ...   
252851  2020-10-31          5     Sat        1      1780      1074      45.1   
252852  2020-10-31          5     Sat        1      1022       532      28.7   
252853  2020-10-31          5     Sat        1      2350      1229      35.0   
252854  2020-10-31          5     Sat        1      1107       717      29.0   
252855  2020-10-31          5     Sat        1       734       399      24.2   

        B01002e2  B01002e3  B02001e2  B02001e3  B02001e5  B11001e1  B11001e2  \
0           20.3      30.1       806       159        22       475       289   
1           43.7      49.6      1069        63        36       525       234   
2           36.9      20.0       975        29         0       315       217   
3           48.3      54.1      1504        24         0       627       273   
4           31.1      32.7      1143       120         7       610       456   
...          ...       ...       ...       ...       ...       ...       ...   
252851      42.6      48.4      1747         0         0       616       514   
252852      28.2      28.9       559        55         0       326       177   
252853      34.6      36.1      1758        20       116       683       471   
252854      28.8      29.3       342        90         4       350       240   
252855      24.5      21.9       336        16         0       194       116   

        B11001e3  EDU_under  EDU_higher  B19013e1    B19025e1  B19301e1  \
0             59   0.561995    0.048518   18972.0  10545000.0    7621.0   
1            153   0.057432    0.247748   2

### 2b 
The 3 COVID related policies I am choosing to focus on are as follows:
* 3/20/20 - the stay-at-home order, which I call 'b_shutdown' and 'a_shutdown' to signify times before and after COVID
* 8/10/20 - more non-essential businesses reopened, 'reopen'
I chose these policies because I figured dailyvisits to the types of businesses would be greatly impacted by the date of shutdown (before and after), and after reopening. 

In [562]:
# creating new variables regarding to Arizona COVID regulations
azMerged['b_shutdown'] = np.where(azMerged['date'] <= '2020-03-20', 1, 0)
azMerged['a_shutdown'] = np.where(azMerged['date'] > '2020-03-20', 1, 0)
azMerged['reopen'] = np.where(azMerged['date'] >= '2020-08-10', 1, 0)

### 2c
I am using a lot of the varibles provided in the class examples plus my three newly created variables. After going through the list of variables, I figured all of the ones below will have some sort of an effect on the dailyvisits to the two business groups I've specified earlier. 

In [563]:
# choosing independent variables
X = azMerged.loc[:,['weekend','region','bizgroup','dayname','month',
                               'B01001e1','B01002e1','B02001e2',
                               'B02001e3','B02001e5','B11001e1',
                               'B11001e2','B11001e3','EDU_under',
                               'EDU_higher','B19013e1','B19025e1',
                               'B19301e1','TAVG','cases','deaths',
                               'b_shutdown','a_shutdown','reopen']]
# defining logged dependent variable 
y = np.log(azMerged.loc[:,'dailyvisits'] + 1)

In [564]:
# showing the data types of each variable
X.dtypes

weekend         int64
region         object
bizgroup       object
dayname        object
month          object
B01001e1        int64
B01002e1      float64
B02001e2        int64
B02001e3        int64
B02001e5        int64
B11001e1        int64
B11001e2        int64
B11001e3        int64
EDU_under     float64
EDU_higher    float64
B19013e1      float64
B19025e1      float64
B19301e1      float64
TAVG          float64
cases         float64
deaths        float64
b_shutdown      int64
a_shutdown      int64
reopen          int64
dtype: object

In [565]:
# get dummies for all categorical variables
X = pd.get_dummies(data = X, drop_first = True)
X.head()

weekend  B01001e1  B01002e1  B02001e2  B02001e3  B02001e5  B11001e1  \
0        0      1605      22.0       806       159        22       475   
1        0      1276      46.4      1069        63        36       525   
2        0      1066      25.5       975        29         0       315   
3        0      1551      48.4      1504        24         0       627   
4        0      2065      31.8      1143       120         7       610   

   B11001e2  B11001e3  EDU_under  EDU_higher  B19013e1    B19025e1  B19301e1  \
0       289        59   0.561995    0.048518   18972.0  10545000.0    7621.0   
1       234       153   0.057432    0.247748   27298.0  22996600.0   19414.0   
2       217        79   0.545966    0.105066   13250.0   8022300.0    8608.0   
3       273       273   0.029752    0.582645   57098.0  54859600.0   38017.0   
4       456       190   0.403846    0.173077   38839.0  27257000.0   15326.0   

   TAVG  cases  deaths  b_shutdown  a_shutdown  reopen  bizgroup_2.0  \
0  50.0    0.0     0.0           1           0       0             1   
1  50.0    0.0     0.0           1           0       0             0   
2  50.0    0.0     0.0           1           0       0             1   
3  50.0    0.0     0.0           1           0       0             1   
4  50.0    0.0     0.0           1           0       0             1   

   dayname_Mon  dayname_Sat  dayname_Sun  dayname_Thur  dayname_Tue  \
0            0            0            0             0            0   
1            0            0            0             0            0   
2            0            0            0             0            0   
3            0            0            0             0            0   
4            0            0            0             0            0   

   dayname_Wed  month_10  month_4  month_7  
0            1         0        0        0  
1            1         0        0        0  
2            1         0        0        0  
3            1         0        0        0  
4            1         0        0        0

In [566]:
# summary of variables 
X.describe()

weekend       B01001e1       B01002e1       B02001e2  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean        0.273779    1632.118773      35.346258    1183.316318   
std         0.445898     731.081105       9.386284     547.308866   
min         0.000000     146.000000      18.000000      48.000000   
25%         0.000000    1137.000000      28.200000     790.000000   
50%         0.000000    1549.000000      33.700000    1112.000000   
75%         1.000000    2022.000000      40.600000    1487.000000   
max         1.000000    6464.000000      71.700000    5875.000000   

            B02001e3       B02001e5       B11001e1       B11001e2  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean      117.009552      55.226711     605.316880     332.259253   
std       159.451133     118.275489     260.044375     174.615229   
min         0.000000       0.000000      43.000000       0.000000   
25%        16.000000       0.000000     427.000000     207.000000   
50%        67.000000      24.000000     571.000000     310.000000   
75%       159.000000      65.000000     700.000000     423.000000   
max      1608.000000    1153.000000    1805.000000    1650.000000   

            B11001e3      EDU_under     EDU_higher       B19013e1  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean      203.519683       0.207258       0.341867   48788.329889   
std       124.993023       0.179559       0.210321   24988.276281   
min         0.000000       0.000000       0.000000    8985.000000   
25%       119.000000       0.050697       0.150376   30250.000000   
50%       186.000000       0.155093       0.319507   43079.000000   
75%       264.000000       0.336673       0.513761   61552.000000   
max      1391.000000       0.755714       0.883257  151042.000000   

           B19025e1       B19301e1           TAVG          cases  \
count  2.470680e+05  247068.000000  247068.000000  247068.000000   
mean   3.898937e+07   26551.310332      77.358581   56863.431833   
std    2.964434e+07   16659.584706      16.965576   61797.626613   
min    2.195100e+06    4199.000000      50.000000       0.000000   
25%    1.939720e+07   12836.000000      60.000000       1.000000   
50%    3.214940e+07   22741.000000      77.000000    3578.000000   
75%    4.949580e+07   34641.000000      90.000000  111446.000000   
max    2.027535e+08   93676.000000     105.000000  158765.000000   

              deaths     b_shutdown     a_shutdown         reopen  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean     1145.009649       0.264996       0.735004       0.217147   
std      1375.266200       0.441332       0.441332       0.412304   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%       137.000000       0.000000       1.000000       0.000000   
75%      1882.000000       1.000000       1.000000       0.000000   
max      3606.000000       1.000000       1.000000       1.000000   

        bizgroup_2.0    dayname_Mon    dayname_Sat    dayname_Sun  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean        0.779457       0.133146       0.140633       0.133146   
std         0.414613       0.339733       0.347644       0.339733   
min         0.000000       0.000000       0.000000       0.000000   
25%         1.000000       0.000000       0.000000       0.000000   
50%         1.000000       0.000000       0.000000       0.000000   
75%         1.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

        dayname_Thur    dayname_Tue    dayname_Wed       month_10  \
count  247068.000000  247068.000000  247068.000000  247068.000000   
mean        0.150493       0.133146       0.158944       0.217147   
std         0.357555       0.339733       0.365625       0.412304   
min         0.

## Question 3
### 3a
I narrowed down the list of variables I will be using in my regression based on the ones I did not think would be useful in predicting y ~ 'log(dailyvisits+1)'. 
After using the train_test_split() command and the MinMaxScalar() function, I found there were 185,301 observations in the training sample and 61,767 in the test sample. There shouldn't be a variable that tells us how the split went because python does this randomly. The train data has 75% of the azMerged and the test data has 25% of azMerged, randomly chosen by python.

In [567]:
# new list of variables to predict log(dailyvisits+1)
features = azMerged.loc[:,['weekend','region','bizgroup','dayname','month',
                               'B01001e1','B01002e1','B11001e1',
                               'B11001e2','B11001e3',
                               'EDU_higher','B19013e1','B19025e1',
                               'B19301e1','TAVG','cases','deaths',
                               'b_shutdown','a_shutdown']]

In [568]:
# get dummies for all categorical variables
features = pd.get_dummies(data = X, drop_first = True)
# features.head()

In [569]:
# seperating Local Merged into a training sample and test sample
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features,y,
                                                    random_state = 1)

In [570]:
# standardizing all the 'features' 
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [571]:
# num of observations in the training sample
len (y_train)

185301

In [572]:
# num of observations in the test sample
len (y_test)

61767

### 3b - Linear Regression
The estimates are in the table below and the R-squared is .131.
##### Statistically significant coefficients 
All variables are statistically significant (p-value<0.05) except for:
* 'cases' the cumulative amount of cases
* 'month_10' the month of October

##### Variables with positive coefficients
* 'B11001e1' - total number of Household
* 'B11001e2' - number of family Households 
* 'B11001e3' - Number of family Households with married couples
* 'B19013e1' - median Household income
* 'B19301e1' - per capita income
* 'TAVG' - average air temperature
* 'deaths' - total cumulative COVID related deaths
* 'b_shutdown' - before COVID shutdown (past - 3/20/20)
* 'a_shutdown' - after COVID shutdown (3/20/20 - present)
* 'dayname_Sat' - Saturday
* 'month_10' - October
* 'month_4' - April
* 'month_7' - July

##### Variables with negative coefficients
* 'weekend' - Saturday and Sunday
* 'B01001e1' - total population
* 'B01002e1' - total population median age
* 'EDU_hugher' - people with a college education and/or higher
* 'B19025e1' - aggregate Household income
* 'cases' - cumulative number of COVID cases 
* 'bizgroup_2.0 - Resturants, Grocery stores and Convenience stores
* 'dayname_Mon' - Monday
* 'dayname_Sun' - Sunday
* 'dayname_Thurs' - Thursday
* 'dayname_Tues' - Tuesday 
* 'dayname_Wed' - Wednesday

The signs of a couple variables surprised me. The ones that were most surprising to me was that the total cumulative COVID related deaths had a positive impact on log(dailyvisits). I was also suprised to see after shutdown had a positive impact on log(dailyvisits).

The R-squared for the test sample is 0.129.

In [573]:
# running linear regression
import statsmodels.api as sm
X2_train = sm.add_constant(X_train)
results = sm.OLS(y_train,X2_train.astype(float)).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            dailyvisits   R-squared:                       0.132
Model:                            OLS   Adj. R-squared:                  0.132
Method:                 Least Squares   F-statistic:                     1081.
Date:                Wed, 14 Apr 2021   Prob (F-statistic):               0.00
Time:                        18:36:56   Log-Likelihood:            -2.5811e+05
No. Observations:              185301   AIC:                         5.163e+05
Df Residuals:                  185274   BIC:                         5.165e+05
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            1.2917      0.023     56.471      0.000       1.247       1.337
weekend         -0.1631      0.005    -33.557      0.000      -0.173      -0.154
B01001e1        -0.0003    1.4e-05    -19.544      0.000      -0.000      -0.000
B01002e1        -0.0052      0.000    -13.892      0.000      -0.006      -0.004
B02001e2      3.134e-05   1.24e-05      2.537      0.011    7.13e-06    5.56e-05
B02001e3      2.281e-05   2.15e-05      1.063      0.288   -1.92e-05    6.49e-05
B02001e5         0.0003   2.43e-05     13.012      0.000       0.000       0.000
B11001e1         0.0007    2.6e-05     26.343      0.000       0.001       0.001
B11001e2         0.0005   4.28e-05     12.333      0.000       0.000       0.001
B11001e3         0.0006    5.8e-05     10.334      0.000       0.000       0.001
EDU_under        0.0314      0.026      1.229      0.219      -0.019       0.082
EDU_higher      -0.4367      0.029    -15.083      0.000      -0.493      -0.380
B19013e1      1.947e-06   2.19e-07      8.874      0.000    1.52e-06    2.38e-06
B19025e1     -4.767e-09   2.85e-10    -16.742      0.000   -5.32e-09   -4.21e-09
B19301e1      5.151e-06    4.8e-07     10.721      0.000    4.21e-06    6.09e-06
TAVG             0.0023      0.000      5.656      0.000       0.001       0.003
cases         -7.45e-07   1.07e-06     -0.694      0.488   -2.85e-06    1.36e-06
deaths           0.0002    5.4e-05      2.825      0.005    4.67e-05       0.000
b_shutdown       1.1106      0.011    102.516      0.000       1.089       1.132
a_shutdown       0.1812      0.016     11.496      0.000       0.150       0.212
reopen           0.0289      0.019      1.497      0.134      -0.009       0.067
bizgroup_2.0    -0.3919      0.006    -70.692      0.000      -0.403      -0.381
dayname_Mon     -0.2121      0.009    -24.672      0.000      -0.229      -0.195
dayname_Sat      0.0174      0.005      3.521      0.000       0.008       0.027
dayname_Sun     -0.1805      0.005    -36.099      0.000      -0.190      -0.171
dayname_Thur    -0.1186      0.008    -14.333      0.000      -0.135      -0.102
dayname_Tue     -0.1754      0.009    -20.542      0.000      -0.192      -0.159
dayname_Wed     -0.1511      0.008    -18.521      0.000      -0.167      -0.135
month_10         0.0289      0.019      1.497      0.134      -0.009       0.067
month_4          0.0739      0.012      6.236      0.000       0.051       0.097
month_7          0.0783      0.025      3.120      0.002       0.029       0.128
==============================================================================
Omnibus:                      541.844   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              422.496
Skew:                          -0.014   Prob(JB):                     1.80e-92
Kurtosis:    

In [574]:
# printing r^2 for train and test sample -- so I don't have to run the full regression again
from sklearn.linear_model import LinearRegression

linreg = LinearRegression().fit(X_train,y_train)

coef = linreg.coef_

print('R-squared score (training): {:.3f}'.format(linreg.score(X_train,y_train)))
print('R-squared score (test): {:.3f}'.format(linreg.score(X_test,y_test)))

R-squared score (training): 0.132
R-squared score (test): 0.129


#### 3c - Lasso Regression
The penalty parameters (alphas) I chose are 0.001, 0.05, 0.1, 1, and 5. 
All of the R-squared values for the training and test samples are in the output below. 

##### Penalty parameter
Among the penaly parameters I chose, the most appropriate one is 0.001. Alpha = 0.001 is the most appropriate for this model because the it is maximizing the testing score alpha and the r-squared for the training sample is not too much larger (overfitting). The r-squared for the test model (0.124) is not too much smaller than the r-squared from the linear regression model, which would indicate underfitting. 

##### Zero and non-zero features
In the Lasso model with the preferred penalty parameter (0.001), there are only 3 'features' whose coefficients that have zero or close to zero coefficients. There are 22 'features' whose coefficients have non-zeor coefficients. 

##### Coefficient signs / magnitude
The only coefficients that were a surprise to me this time around was that of 'weekend'. The coefficient is negative and I would have thought the weekend would cause a greater influx of dailyvisits to stores over the weekdays because more people are off of work on the weekends.
Additionally I was confused as to why the magnitude for average air temp was printed in my optput because the coefficient for it is 0.

##### 5 most important 'features'
* 'B19301e1' - per capita income
* 'region' 
* 'dayname'
* 'cases'
* 'B19025e1' - agg HH income

The order of importance is consistent with my prior. I had a feeling that log(dailyvisits) would be greatly affected by the income of a household and the the measure of the amount a person makes in Arizona. I also figured the day of the week would impact the number of dailyvisits. Additionally, the number of cases, which is negative in this regression is very much consistent with my prior. As the number of cases rise in Arizona, the fewer people will visit stores because more people are getting sick and less willing to make a trip outside the house (ceteris paribus).

In [575]:
for alpha in[0.001,0.05,0.1,1,5]:
    linlasso = Lasso(alpha=alpha,max_iter=10000).fit(X_train_scaled,y_train)
    r2_train=linlasso.score(X_train_scaled,y_train)
    r2_test=linlasso.score(X_test_scaled,y_test)
    
    print('Alpha = {}\nFeatures kept: {}, r-squared training: {:.3f}, \
r-squared test: {:.3f}\n'
          .format(alpha,np.sum(linlasso.coef_!=0),r2_train,r2_test))

Alpha = 0.001
Features kept: 20, r-squared training: 0.127, r-squared test: 0.125

Alpha = 0.05
Features kept: 3, r-squared training: 0.075, r-squared test: 0.074

Alpha = 0.1
Features kept: 2, r-squared training: 0.020, r-squared test: 0.020

Alpha = 1
Features kept: 0, r-squared training: 0.000, r-squared test: -0.000

Alpha = 5
Features kept: 0, r-squared training: 0.000, r-squared test: -0.000



In [576]:
# alpha is the penalty parameter
from sklearn.linear_model import Lasso

# using the desired penalty parameter


linlasso = Lasso(alpha = 0.0001, max_iter = 10000).fit(X_train_scaled,y_train)

print('Non-zero features: {}'.format(np.sum(linlasso.coef_!=0)))
print('R-squared score (training): {:.3f}'.format(linlasso.score(X_train_scaled,y_train)))
print('R-squared score (test): {:.3f}\n'.format(linlasso.score(X_test_scaled,y_test)))
print('Features with non-zero weight (sorted by absolute magnitude):')

for e in sorted (list(zip(list(X),linlasso.coef_)),
                 key = lambda e:-abs(e[1])):
    if e[1]!=0:
        print('\t{}, {:.3f}'.format(e[0],e[1]))

Non-zero features: 24
R-squared score (training): 0.132
R-squared score (test): 0.129

Features with non-zero weight (sorted by absolute magnitude):
	B01001e1, -1.411
	B11001e1, 1.067
	b_shutdown, 0.839
	B11001e2, 0.790
	B11001e3, 0.773
	B19025e1, -0.772
	deaths, 0.406
	bizgroup_2.0, -0.392
	B19301e1, 0.357
	EDU_higher, -0.343
	B02001e5, 0.334
	B01002e1, -0.243
	B19013e1, 0.233
	dayname_Mon, -0.208
	dayname_Sun, -0.197
	dayname_Tue, -0.171
	dayname_Wed, -0.147
	weekend, -0.142
	dayname_Thur, -0.115
	TAVG, 0.114
	B02001e3, -0.015
	month_7, -0.014
	month_4, -0.012
	a_shutdown, -0.000


In [577]:
import eli5
from eli5.sklearn import PermutationImportance

perm = PermutationImportance(linlasso).fit(X_test_scaled,y_test)
eli5.show_weights(perm)

### 3d - Ridge Regression
The penalty parameters (alphas) I chose are 0.001, 0.05, and 1. 
All of the R-squared values for the training and test samples are in the output below. 

##### Penalty parameter
I am noticing that all 3 penalty parameters yeild the same results regarding test and training sample r-squared values. However, I am choosing 1 as the optimal penalty  because I used Gridsearch to chose an alpha that yeilds the highest score and that alpha was 1.389. 

##### Zero and non-zero features
There are 0 features with zero as the coefficient and 25 (all) features with non-zero coefficients. 

##### Coefficient signs / magnitude
I think what surprised me the most was weekend had a negative coefficient again. As I stated previously, I figured more dailyvisits would occur on the weekends. Additionally the coefficient for 'B19013e1' (median hh income is negative. This surprised me because I would have thought that the greater the household income is the greater the log(dailyvisits) would be (ceteris paribus).

The R-squared for the test sample is .129 for the Ridge regression. This is the same as the r-squared for the lasso model but lower than the linear refression model.

##### 5 most important 'features'
* 'region'
* 'dayname'
* 'B19025e1' - agg HH income
* 'TAVG' - average air temp
* 'cases' 

The only comments I have on these most important features is that I am puzzeled how the average air temp can be so important of a feature to this model. Maybe if the air temp is really extreme, that could deter people from leaving the house and decrease log(dailyvisits), however, I don't understand how it is up there in the top 5. 

Ridge identifies 4 of the same most important features as Lasso does. The only difference is that Lasso identifies per capita income as top most important variable while Ridge identifies average air temp. 

In [578]:
# fit the ridge linear regression model
# 0.001 = pentalty parameter
from sklearn.linear_model import Ridge

linridge = Ridge(alpha=0.01).fit(X_train_scaled,y_train)

print('Ridge regression coef:\n{}'.format(linridge.coef_))
print('R-squared score (training): {:.3f}'.format(linridge.score(X_train_scaled,y_train)))
print('R-squared score (test): {:.3f}'.format(linridge.score(X_test_scaled,y_test)))
print('Number of non-zero features: {}'.format(np.sum(linridge.coef_!=0)))

Ridge regression coef:
[-0.163105   -1.72327582 -0.28092037  0.18251949  0.03664225  0.36441932
  1.20868753  0.87095741  0.8344101   0.02374046 -0.38569659  0.27656182
 -0.95595406  0.46084417  0.12473271 -0.11806247  0.5499777   0.35705093
 -0.35705093 -0.07867423 -0.39186105 -0.21208602  0.01739067 -0.18049566
 -0.11858145 -0.17544068 -0.15109192 -0.07867422 -0.14140757 -0.13696913]
R-squared score (training): 0.132
R-squared score (test): 0.129
Number of non-zero features: 30


In [579]:
# 0.01 = pentalty parameter
linridge = Ridge(alpha=0.01).fit(X_train_scaled,y_train)

print('Ridge regression coef:\n{}'.format(linridge.coef_))
print('R-squared score (training): {:.3f}'.format(linridge.score(X_train_scaled,y_train)))
print('R-squared score (test): {:.3f}'.format(linridge.score(X_test_scaled,y_test)))
print('Number of non-zero features: {}'.format(np.sum(linridge.coef_!=0)))

Ridge regression coef:
[-0.163105   -1.72327582 -0.28092037  0.18251949  0.03664225  0.36441932
  1.20868753  0.87095741  0.8344101   0.02374046 -0.38569659  0.27656182
 -0.95595406  0.46084417  0.12473271 -0.11806247  0.5499777   0.35705093
 -0.35705093 -0.07867423 -0.39186105 -0.21208602  0.01739067 -0.18049566
 -0.11858145 -0.17544068 -0.15109192 -0.07867422 -0.14140757 -0.13696913]
R-squared score (training): 0.132
R-squared score (test): 0.129
Number of non-zero features: 30


In [580]:
# 1 = pentalty parameter
linridge = Ridge(alpha=1).fit(X_train_scaled,y_train)

print('Ridge regression coef:\n{}'.format(linridge.coef_))
print('R-squared score (training): {:.3f}'.format(linridge.score(X_train_scaled,y_train)))
print('R-squared score (test): {:.3f}'.format(linridge.score(X_test_scaled,y_test)))
print('Number of non-zero features: {}'.format(np.sum(linridge.coef_!=0)))

Ridge regression coef:
[-0.16309792 -1.70354607 -0.27914109  0.17294858  0.03258129  0.3634143
  1.20058783  0.86562645  0.83099479  0.02210081 -0.38411205  0.27430253
 -0.9477002   0.45707126  0.12531554 -0.09787917  0.5263484   0.35719866
 -0.35719866 -0.07669219 -0.39188491 -0.21214971  0.01740227 -0.18050019
 -0.11851724 -0.17544895 -0.15102907 -0.07669219 -0.14113134 -0.13937514]
R-squared score (training): 0.132
R-squared score (test): 0.129
Number of non-zero features: 30


In [581]:
from sklearn.linear_model  import RidgeCV

n_alphas=50
alphas=np.logspace(-5,2,n_alphas)
linridge_cv=RidgeCV(alphas=alphas).fit(X_train_scaled,y_train)
linridge_cv.alpha_

1.389495494373136

In [582]:
# pip install eli5

In [583]:
perm = PermutationImportance(linridge).fit(X_test_scaled,y_test)
eli5.show_weights(perm)